src/plist.h

```diff
@@ -26,38 +26,85 @@
 #ifndef GAP_PLIST_H
 #define GAP_PLIST_H
 
+#include <src/debug.h> /* for GAP_ASSERT */
 
 /****************************************************************************
 **
-
 *F  NEW_PLIST(<type>,<plen>)  . . . . . . . . . . . allocate a new plain list
 **
 **  'NEW_PLIST'  allocates    a new plain   list  of  type <type> ('T_PLIST',
 **  'T_SET', 'T_VECTOR') that has room for at least <plen> elements.
 **
-**  Note that 'NEW_PLIST' is a  macro, so do not call  it with arguments that
-**  have side effects.
 */
-#define NEW_PLIST(type,plen)            NewBag(type,((plen)+1)*sizeof(Obj))
+static inline Obj NEW_PLIST(UInt type, Int plen)
+{
+    GAP_ASSERT(plen >= 0);
+    return NewBag(type, (plen + 1) * sizeof(Obj));
+}
 
+/****************************************************************************
+**
+*F  IS_PLIST( <list> )  . . . . . . . . . . . check if <list> is a plain list
+*/
+static inline Int IS_PLIST(Obj list)
+{
+    return FIRST_PLIST_TNUM <= TNUM_OBJ(list) &&
+           TNUM_OBJ(list) <= LAST_PLIST_TNUM;
+}
 
 /****************************************************************************
 **
-*F  GROW_PLIST(<list>,<plen>) . . . .  make sure a plain list is large enough
+*F  IS_PLIST_ASSERT_CHECK( <list> )  . . . . .check if <list> is a plain list
 **
-**  'GROW_PLIST' grows  the plain list <list>  if necessary to ensure that it
-**  has room for at least <plen> elements.
+**  This version of IS_PLIST checks if 'COPYING' is set, and allows
+**  T_POSOBJ and T_DATOBJ (for backwards compatability reasons, as PLIST
+**  methods are often used on these types).
```

Comment:
Actually, looking at `FuncSTRONGLY_CONNECTED_COMPONENTS_DIGRAPH`, I now wonder if you didn't dicover a genuine bug in it -- it does `SET_ELM_PLIST(val, x, INTOBJ_INT(n+1));` where `val` is a `T_DATOBJ`. In all other places, though, raw integers are stored into and read from `val`. And here an INTOBJ???

Of course we have no code in GAP itself using or testing this, but various packages do. I wonder if some package authors would be willing to help debug it?